src/drivers/imx/esai.c

```diff
@@ -4,37 +4,456 @@
+	/* TODO */
+	/* ESAI_xFCR_xWA(bits) instead of 16 ^^^^ */
+	/* ESAI_xFCR_xFWM(slots) -- compute slots instead of hardcoding
+	 * 32!
+	 */
+
+	/* Set the clock divider to divide EXTAL by 16 (DIV8 from PSR,
+	 * plus a divide by 2 which is mandatory overall)
+	 */
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xDC_MASK, ESAI_xCCR_xDC(2));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xFP_MASK, ESAI_xCCR_xFP(1));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xPSR_MASK, ESAI_xCCR_xPSR_DIV8);
+
+
+	/* Remove ESAI personal reset */
+	dai_update_bits(dai, REG_ESAI_xCR(0), ESAI_xCR_xPR, 0);
+	dai_update_bits(dai, REG_ESAI_xCR(1), ESAI_xCR_xPR, 0);
+
+	dai_update_bits(dai, REG_ESAI_PRRC, ESAI_PRRC_PDC_MASK,
+			ESAI_PRRC_PDC(ESAI_GPIO));
+	dai_update_bits(dai, REG_ESAI_PCRC, ESAI_PCRC_PC_MASK,
+			ESAI_PCRC_PC(ESAI_GPIO));
+
+	tracev_esai("ESAI_REGS_DUMP in esai_set_config");
+	esai_regs_dump(dai);
 	return 0;
 }
 
+static void esai_start(struct dai *dai, int direction)
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
+	/* FIFO enable */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFEN_MASK,
+			ESAI_xFCR_xFEN);
+
+	/* Write a few zero samples, one per channel, for initialization */
+	/* TODO determine the correct number of samples, assuming 2 */
+	if (direction) {
+		dai_write(dai, REG_ESAI_ETDR, 0);
+		dai_write(dai, REG_ESAI_ETDR, 0);
+	}
+
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK,
+			direction ? ESAI_xCR_TE(1) : ESAI_xCR_RE(1));
+
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK,
+			ESAI_xSMB_xS(0x3));
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK,
+			ESAI_xSMA_xS(0x3));
+
+	tracev_esai("ESAI_REGS_DUMP in esai_start");
+	esai_regs_dump(dai);
+}
+
+static void esai_stop(struct dai *dai, int direction)
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
+
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK, 0);
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK, 0);
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK, 0);
+
+	/* disable and reset FIFO */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction),
+			ESAI_xFCR_xFR | ESAI_xFCR_xFEN, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFR, 0);
+
+	tracev_esai("ESAI_REGS_DUMP in esai_stop");
+	esai_regs_dump(dai);
+}
+
 static int esai_trigger(struct dai *dai, int cmd, int direction)
 {
+	tracev_esai("ESAI: trigger");
+	switch (direction) {
+	case DAI_DIR_PLAYBACK:
+		tracev_esai("ESAI: playback trigger");
+		break;
+	case DAI_DIR_CAPTURE:
+		tracev_esai("ESAI: capture trigger");
+		break;
+	default:
+		trace_esai_error("ESAI: <INVALID DIRECTION %d> trigger", direction);
+	}
+
+	switch (cmd) {
+#define CASE(c) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #c ); break;
+#define CASE2(c, p) case COMP_ ## c: tracev_esai("ESAI: trigger cmd DAI_" #p); break;
```

Comment:
Oh...

Reply:
Looks like a lot of stuff left over from previous attempts at getting it to work. Will clean it up tomorrow. This one in particular actually goes away.

Reply:
Done.